core/src/page.rs

```diff
@@ -0,0 +1,379 @@
+                );
+
+                self.location = CursorLocation::Page(parent_page);
+            }
+        }
+    }
+
+    /// Traverse into the two locations below the current node.
+    ///
+    /// Provide a closure which peeks at both values and informs the node which path to take:
+    /// `false` is left and `true` is right.
+    ///
+    /// It is the caller's responsibility to ensure that the current node is an internal node. If
+    /// it is not, then the cursor will traverse into the empty locations below without complaining.
+    ///
+    /// This may attempt to load a page and fail, if the current location is at the end of the
+    /// current page.
+    pub fn traverse_children(
+        &mut self,
+        f: impl FnOnce(&Node, &Node) -> bool,
+    ) -> Result<(), MissingPage> {
+        let (left_idx, right_idx) = if self.depth as usize % DEPTH == 0 {
+            // attempt to load next page if we are at the end of our previous page or the root.
+            let page_id = match self.location {
+                CursorLocation::Root => ROOT_PAGE_ID,
+                CursorLocation::Page(p) => {
+                    let child_page_idx = last_page_path(&self.path, self.depth).load_be::<u8>();
+                    child_page_id(*p.id(), child_page_idx)
+                }
+            };
+
+            let page = self.pages.get(&page_id).ok_or(MissingPage)?;
+            self.location = CursorLocation::Page(page);
+
+            child_node_indices(BitSlice::empty())
+        } else {
+            let path = last_page_path(&self.path, self.depth);
+            child_node_indices(path)
+        };
+
+        // invariant: cursor location is always `Page` at this point onwards.
+        let bit = match self.location {
+            CursorLocation::Root => panic!("Root means depth = 0, we loaded next page; qed"),
+            CursorLocation::Page(p) => f(&p.nodes()[left_idx], &p.nodes()[right_idx]),
+        };
+
+        // Update the cursor's lookup path.
+        self.path
+            .view_bits_mut::<Msb0>()
+            .set(self.depth as usize, bit);
+        self.depth += 1;
+
+        Ok(())
+    }
+}
+
+// extract the relevant portion of the key path to the last page. panics on empty path.
+fn last_page_path(total_path: &KeyPath, total_depth: u8) -> &BitSlice<u8, Msb0> {
+    let prev_page_end = ((total_depth as usize - 1) / DEPTH) * DEPTH;
+    &total_path.view_bits::<Msb0>()[prev_page_end..total_depth as usize]
+}
+
+// Transform a bit-path to an index in a page.
+//
+// The expected length of the page path is between 1 and `DEPTH`, inclusive. A length of 0 returns
+// 0 and all bits beyond `DEPTH` are ignored.
+fn node_index(page_path: &BitSlice<u8, Msb0>) -> usize {
+    let len = page_path.len();
+    if len == 0 {
+        0
+    } else if len > DEPTH {
+        node_index(&page_path[..DEPTH])
+    } else {
+        let i = page_path.load_be::<u8>() as usize;
+        let base = (1 << len) - 2;
+        base + i
```

Comment:
I would move the `-2` here to emphasize that it is the 'conversion' between the number rule in the binary tree and how they are indexed in memory. I had to spend some time before figuring it out

Reply:
It's different than my mental model, which is that there is an offset for each length and an offset within each length - `base + i`, where `base = 2^len - 2` and `i` is the bit-string interpreted as a number.